### DISCUSSION TITLE
Eigenvalue Decomposition

### DISCUSSION THREAD

**Discussion by zoeschindler** [r-lidar/lidR#426] [General]
Hello,

following this question on stackexchange: [https://gis.stackexchange.com/questions/395916/get-eigenvalues-of-large-point-cloud-using-lidr](https://gis.stackexchange.com/questions/395916/get-eigenvalues-of-large-point-cloud-using-lidr) I wanted to open this discussion to keep notified on the implementation of the `eigenvalue_decomposition()` code.
Thanks again for the code!

**Comment by Jean-Romain** [comment]
I'm not sure of the way I want to provide access to such feature. I can see several options:

1. A function `eigen_metrics()` that does exactly the same as `point_metrics(las, .stdshapemetrics)` but more efficiently
2. A transparent way to use internal optimization when running `point_metrics(las, .stdshapemetrics)` 
3. Both 1 and 2
4. A function `eigen_metrics()` with some specific capabilities (which ones)
5. Other

**Comment by bi0m3trics** [comment]
Not that anyone asked, but I'd vote for #2, maybe with an eigen metrics example in the book showing the reader how the functionality (with Rcpp) in lidR already allows them to build efficient functions, but lets the reader know that ensuring they're functions are thread safe and parallelizable is still up to them. It could also serve to provide some further guidance on indexing, ensuring that multiple copies aren't created in memory. This could be done where the actual example answers this exact question but doesn't require @Jean-Romain to account for all possible scenarios. 

Basically, document the things you've taught me!  ;)

**Reply by Jean-Romain** [reply]
Well actually after thinking about it some options are not viable because they are not consistent with the current framework. The function we are talking about will be natively multithreaded with OpenMP. The current way to know if a something is multithreaded in lidR is to used `is.parallelized()` (or display the algorithm)

```r
lidR::shp_plane()
#> Object of class lidR algorithm
#> Algorithm for: shape segmentation 
#> Designed to be used with: segment_shapes 
#> Native C++ parallelization: yes 
#> Parameters: 
#>  - k = 8 <numeric>
#>  - th1 = 25 <numeric>
#> - th2 = 6 <numeric>
```

I cannot make `point_metrics()` single threaded except in some specific case internally handled. Yes indeed I can document it but if I start working this way it's gonna be a mess.

A consistent way could be

```r
eigen_decomposition(las, eigen_values(k = 10))
eigen_decomposition(las, eigen_vectors(k = 10))
eigen_decomposition(las, eigen_metrics(k = 10))
eigen_decomposition(las, eigen_all(k = 10))
```

I don't like it. Maybe

```r
eigen_decomposition(las, eigen_metrics(values = T, vectors = T, metrics = T, k = 10))

eigen_metrics()
#> Object of class lidR algorithm
#> Algorithm for: eigen decomposition 
#> Designed to be used with: eigen_decomposition 
#> Native C++ parallelization: yes 
#> Parameters: 
#> values = TRUE <logical>
#> vectors = TRUE  <logical>
#> metrics = TRUE  <logical>
#>  - k = 10 <numeric>
```

**Comment by Jean-Romain** [comment]
New function `point_eigenvalue` in 3.2.0
